assignments/assignment04.rb

```diff
@@ -0,0 +1,174 @@
+# ========================================================================================
+# Assignment 4
+# ========================================================================================
+
+# ========================================================================================
+#  Problem 1 - Fibonacci
+
+def fib(n)
+  if (0..1).include? n
+  (fib(n-1) + fib(n-2)) 
+end
+
+# expected behavior:
+fib(0)     #=> 1
+fib(1)     #=> 1
+fib(5)     #=> 8
+fib(4)     #=> 5
+fib(12)    #=> 233
+
+
+# ========================================================================================
+#  Problem 2 - Queue
+
+# implement a Queue class that does not use Array.
+
+# expected behavior:
+q = Queue.new
+q.empty?            #=> true
+q.enqueue "first"
+q.empty?            #=> false
+q.enqueue "second"
+q.dequeue           #=> "first"
+q.dequeue           #=> "second"
+q.dequeue           #=> nil
+
+class Queue
+
+  class Node
+    attr: :item, :link
+    def initialize(item, link)
+      @item = item
+      @link = link
+    end
+  end
+
+  def initialize
+    @nodes = nil
+  end
+
+  def enqueue(item)
+    node = @nodes
+    if node
+      while node.link
+        node = node.link
+      end
+      node.link = Node.new item, nil
+    else
+      @nodes = Node.new item, nil
+    end
+  end
+
+  def dequeue
+    node = @nodes
+    if @nodes = node.nil?
+      node.link
+    else
+      node.item
+    end
+  end
+
+  def empty?
+    @nodes.nil?
+  end
+
+  def peek
+    if @nodes.nil?
+      nil
+    else
+      @nodes.item
+    end
+  end
+
+  def length
+    node = @nodes
+    count = 0
+    while node
+      node = node.link
+      count += 1
+      end
```

Comment:
You have an extraneous `end` in here, which causes a syntax error for the rest of the code.